sim/cpp/test_ip_dhcp.cc

```diff
@@ -0,0 +1,291 @@
+        Addr IP_FIN = Addr(IP_BASE.value + POOL_SIZE - 1);
+        Addr IP_OOB = Addr(IP_BASE.value + POOL_SIZE);
+        CHECK(address_pool.addr2idx(IP_BASE) == 0);
+        CHECK(address_pool.addr2idx(IP_FIN) == POOL_SIZE-1);
+        CHECK(address_pool.contains(IP_BASE));
+        CHECK(address_pool.contains(IP_FIN));
+        CHECK_FALSE(address_pool.contains(IP_OOB));
+    }
+
+    // Static IP-address.
+    SECTION("static") {
+        // Set static configuration.
+        client.release(IP_STATIC);
+        timer.sim_wait(10000);
+
+        // Confirm the system never claimed a lease.
+        CHECK(check_leases(server, 0));
+        CHECK(client.state() == DhcpState::STOPPED);
+        CHECK(client.status() == 0);
+        CHECK(net_client.ipaddr() == IP_STATIC);
+    }
+
+    // Information request.
+    SECTION("inform") {
+        // Configure DHCP server metadata.
+        server.set_dns(IP_GOOGLE);
+        server.set_domain("satcat5");
+        server.set_gateway({IP_ROUTER, MASK_24});
+
+        // Make the request.
+        client.inform(IP_STATIC);
+        timer.sim_wait(10000);
+
+        // Confirm the system never claimed a lease.
+        CHECK(check_leases(server, 0));
+        CHECK(client.state() == DhcpState::STOPPED);
+        CHECK(client.status() == 0);
+        CHECK(net_client.ipaddr() == IP_STATIC);
+
+        // Confirm the local subnet is configured.
+        CHECK(net_client.m_ip.route_lookup(IP_BASE) == IP_BASE);
+        CHECK(net_client.m_ip.route_lookup(IP_GOOGLE) == IP_ROUTER);
+    }
+
+    // Manual release / renew cycle.
+    SECTION("renew") {
+        // Release from the INIT state.
+        client.release();
+        timer.sim_wait(5000);
+        CHECK(check_leases(server, 0));
+        CHECK(client.state() == DhcpState::STOPPED);
+        CHECK(client.status() == 0);
+        CHECK(net_client.ipaddr() == ADDR_NONE);
+
+        // Renew from the INIT state.
+        CHECK(client.status() == 0);
+        client.renew();
+        timer.sim_wait(5000);
+        CHECK(check_leases(server, 1));
+        CHECK(client.state() == DhcpState::BOUND);
+        CHECK(client.status() > 0);
+        CHECK(net_client.ipaddr() == IP_BASE);
+
+        // Renew from the BOUND state (should reuse).
+        client.renew();
+        CHECK(client.state() == DhcpState::RENEWING);
+        timer.sim_wait(5000);
+        CHECK(check_leases(server, 1));
+        CHECK(client.state() == DhcpState::BOUND);
+        CHECK(client.status() > 0);
+        CHECK(net_client.ipaddr() == IP_BASE);
+
+        // Release from the BOUND state.
+        client.release();
+        timer.sim_wait(5000);
+        CHECK(check_leases(server, 0));
+        CHECK(client.state() == DhcpState::STOPPED);
+        CHECK(client.status() == 0);
+        CHECK(net_client.ipaddr() == ADDR_NONE);
+    }
+
+    // Renew but the original lease is invalid.
+    SECTION("renew2") {
+        // Get the initial lease.
+        timer.sim_wait(10000);
```

Comment:
kind of seems like these timeouts should be defined somewhere

Reply:
RFC2131 requires certain timeouts exist but doesn't specify values.  So there's no specific standards document to link to.  Here, 10,000 msec = 10 sec is simply shorthand for "a long time" and the exact value is not critical.  By design, the requirements set for each test are not linked to the constants defined in ip_dhcp.cc.